Pubdev 3694 pca wide data

wendycwong opened the pull request:
For a dataset A say, dimension is m by n where m is the number of samples and n is the feature space dimension.  Let k be the number of eigenvectors we want to use.  The eigenvector matrix will be stored in a double array of size n by k.  This works when n is small.  However, when n is huge compared to m, we get an out of memory error using GramSVD as demonstrated by dataset from Erin in the JIRA.

My fix: instead of generating SVD for transpose(A)*A, we generate SVD for A*transpose(A).  This will result in a much smaller eigenvector matrix of size m by k.  Then, you get back the original eigenvector matrix by transpose(A) * U where U is the eigenvector matrix.  The devil is in the details of course.

Tests:  JUNIt, RUnit and PyUnit tests are written to test for accuracy of implementation.

JUNIt test in PCATest.java:
1.  testWideDataSedtsSmallDataNumeric(): test PCA results on small numeric data columns with original PCA and wide dataset PCA.  The eigenvalues/eigenvectors agree well with each other.
2. testWideDataSedtsSmallData(): test PCA results on small datasets with both categorical and numeric data columns.  There are more difference between the two sets of eigenvalues/eigenvectors.  In this case, I check the reconstructed datasets with the original dataset and see if they agree.  Both methods provide good reconstructed datasets that agrees with the original datasets well.
3. testWideDataSets() and testWideDataSetsWithNAs(): same test as in 2 but with a bigger dataset and with NAs in them to make sure it works.

RUnit test: in runit_pca_pubdev_3694_Rotterdam_large.R derived from Erin JIRA.  Here we run and compare the PCA from R and H2O.  We compare the errors made in the reconstructed datasets and as long as they are close, our PCA is working.

Python test: in pyunit_pubdev_3694_Rotterdam_large.py derived from Erin JIRA.  Simple test to just call and run PCA with the fat dataset and make sure it run to completion.

tomasnykodym commented on h2o-algos/src/test/java/hex/pca/PCATest.java:
```diff
@@ -105,6 +103,207 @@
+      // check to make sure eigenvalues and eigenvectors are the same
+      // compare eigenvectors and eigenvalues generated by original PCA and wide dataset PCA.
+      TestUtil.checkStddev(modelW._output._std_deviation, modelN._output._std_deviation, 1.5);
+      boolean[] flippedEig = TestUtil.checkEigvec( modelW._output._eigenvectors, modelN._output._eigenvectors, 2);
```
2 seems to be suspiciously high tolerance. I agree that if the projections match it should be ok, but just to be sure, maybe we should double check by implementing the same method in R (should be only few lines of code in R) and comparing the results.

wendycwong replied:
Good point here.  I will do the following instead:
1. Implement the code in octave and obtain the eigenvalues, eigenvectors;
2. Compare the eigenvalues, eigenvectors from Octave with our H2O implementation.

tomasnykodym commented on h2o-algos/src/main/java/hex/gram/Gram.java:
```diff
@@ -667,6 +669,28 @@ public final void addRow(DataInfo.Row row, double w) {
+    }
+
+    // cat X cat
+    if (catOffsets != null) {
```
I don't understand this code. Are we doing dot product of the rows? If so the categorical part is a dot product of two spare binary vectors. Meaning which should just sum up the ids appearing in both? in that case, we need to do a variation of a sorted merge, i.e. need two indexes and only advance i if rowj.binIds[j] is bigger and vice versa. Why are we adding the cardinality? Or do I get it wrong ?

Also, if we're doing dot product of the two rows, I'd pu it into a Row.dot(Row) method.

wendycwong replied:
Tomas: 

That is a bug in the categorical data treatment on my part.  Thanks!  After I fixed it, the test results are much better.  The eigenvectors/eigenvalues match exactly using original PCA and my side dataset PCA.  Thanks!

I am doing a dot product between rows and will adopt your proposal.  Thanks, again.  Wendy

tomasnykodym commented on h2o-algos/src/main/java/hex/gram/Gram.java:
```diff
@@ -731,6 +755,82 @@ public void mul(double [] x, double [] res){
+            _dinfo.extractDenseRow(chks, j, rowj);
+
+            if ((!rowi.isBad() && rowi.weight != 0) && (!rowj.isBad() && rowj.weight != 0)) {
+              processRow(rowi, rowj, rowIOffset, j + rowOffset);
```
Why process row? You can do this directly here. If you override map() like this there is no need to extend a FrameTask2, you can extend MRTask directly and loose the processRow.

wendycwong replied:
This is true.  I copied operation from the Gram class since I will in the end obtain a symmetric matrix and use the same scaling of 1/n.

wendycwong commented:
Tomas:  Moved OuterGram to inherit from MRTask.  Added row.dotSame method to carry out the dot product of two rows with the same column types or from the same dataset.   Thanks, Wendy

tomasnykodym commented on h2o-algos/src/main/java/hex/DataInfo.java:
```diff
@@ -856,6 +856,30 @@ public void addNum(int id, double val) {
+      double elementij = 0.0;
+      double weights = this.weight*rowj.weight;
+      for(int i = 0; i < this.nNums; ++i)  {
+        elementij += weights*this.numVals[i]*rowj.numVals[i]; // multiply numerical parts of columns
```
minor thing, I'd move the weight multiplication to the very end, to save an op per element

wendycwong replied:
Good point!

tomasnykodym commented on h2o-algos/src/main/java/hex/glrm/GLRM.java:
```diff
@@ -70,7 +70,10 @@
-    long mem_usage = 8 /*doubles*/ * (_parms._k * _train.numCols() + _parms._k );  // loose estimation of memory usage
+    HeartBeat hb = H2O.SELF._heartbeat;
+    double p = hex.util.LinearAlgebraUtils.numColsExp(_train,true);
+    long mem_usage = (long)(hb._cpus_allowed * p*_parms._k * 8 * Math.log((double)_train.lastVec().nChunks())/
```
seems strange to be taking number of cpus from heartbeat. I think it's a member on H2O object or somewhere similar.

wendycwong replied:
Tomas:  I looked at the other checkMemoryFootPrint() in deep learning and elsewhere, they all use the H2O.SELF._heartbeat call and then you can call get_free_mem() or other functions to get what you want.  Please advice on proper use because I have no idea.

tomasnykodym commented on h2o-algos/src/test/java/hex/DTMatrix/DMatrixTest.java:
```diff
@@ -0,0 +1,276 @@
+
+            // check to make sure eigenvalues and eigenvectors are the same
+            // compare eigenvectors and eigenvalues generated by original PCA and wide dataset PCA.
+            TestUtil.checkStddev(modelW._output._std_deviation, modelN._output._std_deviation, 1e-2);
```
1e-2 seems high tolerance, especially since without nas you have 1e-12. Seems suspicious.

wendycwong replied:
I thought I changed them all.  Looks like a git mess up.  Checked my code and they have been changed.  Let's try this again.

tomasnykodym commented on h2o-core/src/main/java/water/rapids/ast/prims/reducers/AstNaRowCnt.java:
```diff
@@ -0,0 +1,50 @@
+
+  @Override
+  public Val apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
+    Frame fr = stk.track(asts[1].exec(env)).getFrame();
```
This is really slow and will pull data on a single node. Might be ok for datasets with small number of rows but once you add it someone will use it on a big one some item in the future and get OOM.  
I'd prefer if you can do an MRTask here (the logic is the same, you just do it per-chunk and sum the results up in reduce)..

wendycwong replied:
Tomas:

I have abandoned this route.  Will remove them from codebase here.   Thanks,
Wendy

wendycwong commented:
@h2o-ops please test!

wendycwong commented:
All tests passed on my personal Jenkins branch.
